### PR TITLE
feature: added test cases and supportd set retries. 

### PR DESCRIPTION
**Pull request by moonming**

### CODE REVIEW

File: lib/resty/shdict/simple.lua [hunk 1]
```diff
@@ -21,6 +21,7 @@ function _M.gen_shdict_methods (opts)
     local neg_ttl = opts.negative_ttl
     local disable_shdict = opts.disable_shdict
     local dict_name = opts.dict_name
+    local set_retries = opts.set_retries or 1
```
Comment (membphis) [inline lib/resty/shdict/simple.lua]: how about the name `force_retries`?

Comment (moonming) [reply lib/resty/shdict/simple.lua]: @membphis fixed

File: lib/resty/shdict/simple.lua [hunk 2]
```diff
@@ -54,7 +55,14 @@ function _M.gen_shdict_methods (opts)
                 end
 
                 -- with force:
-                ok, err = shdict:set(key, value, ttl)
+                local retries = 0
+                while not ok and force_retries > retries do
```
Comment (agentzh) [inline lib/resty/shdict/simple.lua:59]: It's more readable to put variables on the LHS while the constants on the RHS.

File: lib/resty/shdict/simple.lua [hunk 3]
```diff
@@ -54,7 +55,14 @@ function _M.gen_shdict_methods (opts)
                 end
 
                 -- with force:
-                ok, err = shdict:set(key, value, ttl)
```
Comment (agentzh) [inline lib/resty/shdict/simple.lua:57]: We should avoid using "safe_set" above on line 49.

File: lib/resty/shdict/simple.lua [hunk 4]
```diff
@@ -54,7 +55,14 @@ function _M.gen_shdict_methods (opts)
...
+                local retries = 0
```
Comment (agentzh) [inline lib/resty/shdict/simple.lua:58]: This variable name looks confusing. We are actually recording the number of "tries" in this variable, not "retries".

File: lib/resty/shdict/simple.lua [hunk 5]
```diff
@@ -21,6 +21,7 @@ function _M.gen_shdict_methods (opts)
     local neg_ttl = opts.negative_ttl
     local disable_shdict = opts.disable_shdict
     local dict_name = opts.dict_name
+    local force_retries = opts.force_retries or 1
```
Comment (agentzh) [inline lib/resty/shdict/simple.lua:24]: "force_retries" is a bad name. Maybe you mean "max_retries" or "max_tries"? Depending on whether you count the final (successful) try.

File: t/shdict-simple.t [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    }
+--- request
+GET /t
+--- response_body
+true
+true
+true
+true
+--- error_log
+try to set by force 1th time
```
Comment (agentzh) [inline t/shdict-simple.t:121]: We should use the wording "retry" not "set by force".

File: lib/resty/shdict/simple.lua [hunk 7]
```diff
@@ -54,7 +55,14 @@ function _M.gen_shdict_methods (opts)
...
 
                 -- with force:
-                ok, err = shdict:set(key, value, ttl)
+                local retries = 0
+                while not ok and force_retries > retries do
+                    ok, err = shdict:set(key, value, ttl)
+                    retries = retries + 1
+                    if DEBUG then
+                        dlog(ctx, 'try to set by force ' .. tostring(retries) .. 'th time')
+                    end
```
Comment (agentzh) [inline lib/resty/shdict/simple.lua:64]: We should only retry when the `err` is "no memory", not any other errors.

Comment (moonming) [reply lib/resty/shdict/simple.lua:64]: fixed by https://github.com/openresty/lua-resty-shdict-simple/pull/4 @agentzh

**Pull request merged**